### PR TITLE
UAR-831 Restructure admin and delegate tables

### PR DESCRIPTION
**Pull request by livzorn**
### JIRA link
_HEEDLS-831_

### Description
Make adjustments to the AdminUsers and Candidates tables to accommodate the new Users table.

-----
### Developer checks
(Leave tasks unticked if they haven't been appropriate for your ticket.)

I have:
- [x] Run the formatter and made sure there are no IDE errors.
- [ ] Written tests for the changes (accessibility tests, unit tests for controller, data services, services, view models, etc)
- [x] Manually tested my work with and without JavaScript. Full manual testing guidelines can be found here: https://softwiretech.atlassian.net/wiki/spaces/HEE/pages/6703648740/Testing
- [ ] Updated/added documentation in Swiki and/or Readme. Links (if any) are below:
  - [doc_1_here](link_1_here)
- [ ] Updated my Jira ticket with information about other parts of the system that were touched as part of the MR and have to be sanity tested to ensure nothing’s broken.
- [x] Scanned over my own MR to ensure everything is as expected.

### CODE REVIEW

File: DigitalLearningSolutions.Data.Migrations/202203161448_RestructureAdminAndDelegateTables.cs [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+            Delete.Index("IX_Candidates_CentreID_FirstName").OnTable("DelegateAccounts");
+            Delete.Index("IX_Candidates_CentreID").OnTable("DelegateAccounts");
+            Alter.Column("CandidateNumber").OnTable("DelegateAccounts").AsString(20).NotNullable();
+            Create.Index("IX_Candidates_CandidateNumber").OnTable("DelegateAccounts").OnColumn("CandidateNumber")
+                .Ascending().WithOptions().Unique().WithOptions().NonClustered();
+            Create.Index("IX_Candidates_CentreID_FirstName_LastName").OnTable("DelegateAccounts").OnColumn("CentreID")
+                .Ascending()
+                .OnColumn("FirstName").Ascending().OnColumn("LastName").Ascending().WithOptions().NonClustered();
+            Create.Index("IX_Candidates_CentreID_LastName").OnTable("DelegateAccounts").OnColumn("CentreID").Ascending()
+                .OnColumn("LastName").Ascending().WithOptions().NonClustered();
```
Comment (AlexJacksonDS) [inline DigitalLearningSolutions.Data.Migrations/202203161448_RestructureAdminAndDelegateTables.cs:65]: These indexes will reference deprecated columns, do we want this?

File: DigitalLearningSolutions.Data.Migrations/202203161448_RestructureAdminAndDelegateTables.cs [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+            Create.Index("IX_Candidates_Active_CentreID").OnTable("DelegateAccounts").OnColumn("Active").Ascending()
+                .OnColumn("CentreID").Ascending().WithOptions().NonClustered();
+            Create.Index("IX_Candidates_CentreID_DateRegistered").OnTable("DelegateAccounts").OnColumn("CentreID")
+                .Ascending()
+                .OnColumn("DateRegistered").Ascending().WithOptions().NonClustered();
+            Create.Index("IX_Candidates_CentreID_FirstName").OnTable("DelegateAccounts").OnColumn("CentreID")
+                .Ascending()
+                .OnColumn("FirstName").Ascending().WithOptions().NonClustered();
```
Comment (AlexJacksonDS) [inline DigitalLearningSolutions.Data.Migrations/202203161448_RestructureAdminAndDelegateTables.cs:73]: And this one

**Review by AlexJacksonDS — CHANGES_REQUESTED** [review]
Couple of questions about indexes, and you've missed a requirement:

- Create AdminUsers and Candidates views to display data from Users, AdminAccounts and DelegateAccounts as it was being displayed by the old AdminUsers and Candidates tables.
  - Note that we can assume these views will only return entities that have a respective entry in Users - once the data is migrated, the FK(->Users) column on AdminAccounts and DelegateAccounts will be non-null

File: DigitalLearningSolutions.Data.Migrations/202203161448_RestructureAdminAndDelegateTables.cs (hunk 1 above)
Comment (livzorn) [reply DigitalLearningSolutions.Data.Migrations/202203161448_RestructureAdminAndDelegateTables.cs]: Steve has confirmed that we don't need to add back any indexes based on deprecated columns

File: DigitalLearningSolutions.Data.Migrations/202203161448_RestructureAdminAndDelegateTables.cs (hunk 2 above)
Comment (livzorn) [reply DigitalLearningSolutions.Data.Migrations/202203161448_RestructureAdminAndDelegateTables.cs]: gone

**Comment by livzorn** [comment]
Added the AdminUsers and Candidates views

**Review by AlexJacksonDS — APPROVED** [review]

File: DigitalLearningSolutions.Data.Migrations/202203161448_RestructureAdminAndDelegateTables.cs [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+            Rename.Column("ProfileImage").OnTable("AdminAccounts").To("ProfileImage_deprecated");
+            Rename.Column("SkypeHandle").OnTable("AdminAccounts").To("SkypeHandle_deprecated");
+            Rename.Column("PublicSkypeLink").OnTable("AdminAccounts").To("PublicSkypeLink_deprecated");
+            Rename.Column("ResetPasswordID").OnTable("AdminAccounts").To("ResetPasswordID_deprecated");
+
+            Rename.Table("Candidates").To("DelegateAccounts");
+            Alter.Table("DelegateAccounts").AddColumn("UserID").AsInt32().Nullable().ForeignKey("Users", "ID");
+            Alter.Column("DateRegistered").OnTable("DelegateAccounts").AsDateTime().Nullable();
+
+            Delete.Index("IX_Candidates_CandidateNumber").OnTable("DelegateAccounts");
```
Comment (DanBloxham-sw) [inline DigitalLearningSolutions.Data.Migrations/202203161448_RestructureAdminAndDelegateTables.cs:52]: We are deleting 7 indices here, but leaving 3. What is the reasoning behind the choices?

File: DigitalLearningSolutions.Data.Migrations/Scripts/UAR-831-CreateViewsForAdminUsersAndCandidatesTables-UP.sql [hunk 4]
```diff
@@ -0,0 +1,80 @@
+﻿CREATE VIEW AdminUsers AS
+SELECT dbo.Users.ID                                      AS AdminID,
```
Comment (DanBloxham-sw) [inline DigitalLearningSolutions.Data.Migrations/Scripts/UAR-831-CreateViewsForAdminUsersAndCandidatesTables-UP.sql:2]: Should this be the AdminAccounts.AdminID?

File: DigitalLearningSolutions.Data.Migrations/Scripts/UAR-831-CreateViewsForAdminUsersAndCandidatesTables-UP.sql [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+       dbo.Users.ResetPasswordID,
+       dbo.AdminAccounts.IsFrameworkContributor,
+       dbo.AdminAccounts.IsWorkforceManager,
+       dbo.AdminAccounts.IsWorkforceContributor,
+       dbo.AdminAccounts.IsLocalWorkforceManager,
+       dbo.AdminAccounts.IsNominatedSupervisor           AS NominatedSupervisor
+FROM dbo.Users
+         INNER JOIN
+     dbo.AdminAccounts ON dbo.Users.ID = dbo.AdminAccounts.UserID
+WHERE dbo.AdminAccounts.UserID <> NULL
```
Comment (DanBloxham-sw) [inline DigitalLearningSolutions.Data.Migrations/Scripts/UAR-831-CreateViewsForAdminUsersAndCandidatesTables-UP.sql:42]: Don't think we need the WHERE here (or on the Candidates view) as the INNER JOIN should work that magic.

File: DigitalLearningSolutions.Data.Migrations/Scripts/UAR-831-CreateViewsForAdminUsersAndCandidatesTables-UP.sql [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+       dbo.AdminAccounts.Login_deprecated                AS Login,
+       dbo.Users.PasswordHash                            AS Password,
+       dbo.AdminAccounts.CentreID,
+       dbo.AdminAccounts.IsCentreAdmin                   AS CentreAdmin,
+       dbo.AdminAccounts.ConfigAdmin_deprecated          AS ConfigAdmin,
+       dbo.AdminAccounts.IsReportsViewer                 AS SummaryReports,
+       dbo.AdminAccounts.IsSuperAdmin                    AS UserAdmin,
+       dbo.Users.FirstName                               AS Forename,
+       dbo.Users.LastName                                AS Surname,
+       dbo.AdminAccounts.Email,
```
Comment (DanBloxham-sw) [inline DigitalLearningSolutions.Data.Migrations/Scripts/UAR-831-CreateViewsForAdminUsersAndCandidatesTables-UP.sql:12]: Should be the email on the new Users table?

**Review by DanBloxham-sw — CHANGES_REQUESTED** [review]
A few things I've spotted, some changes may be unnecessary.

File: DigitalLearningSolutions.Data.Migrations/202203161448_RestructureAdminAndDelegateTables.cs [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+                .Ascending().WithOptions().Unique().WithOptions().NonClustered();
+            Create.Index("IX_Candidates_Active_CentreID").OnTable("DelegateAccounts").OnColumn("Active").Ascending()
+                .OnColumn("CentreID").Ascending().WithOptions().NonClustered();
+            Create.Index("IX_Candidates_CentreID_DateRegistered").OnTable("DelegateAccounts").OnColumn("CentreID")
+                .Ascending()
+                .OnColumn("DateRegistered").Ascending().WithOptions().NonClustered();
+            Create.Index("IX_Candidates_CentreID").OnTable("DelegateAccounts").OnColumn("CentreID").Ascending()
+                .WithOptions().NonClustered();
+
+            Rename.Column("EmailAddress").OnTable("DelegateAccounts").To("Email");
```
[the remark this answers is not in the document]
Comment (SteveJacksonSoft) [reply DigitalLearningSolutions.Data.Migrations/202203161448_RestructureAdminAndDelegateTables.cs:70]: *Specific to the centre, not unique. Sorry, got the wrong end of the stick. It's possible we should add a uniqueness index on (email, centreId), but let's check with Kevin. @livzorn would you do the honours?

[the remark this answers is not in the document]
Comment (SteveJacksonSoft) [reply DigitalLearningSolutions.Data.Migrations/202203161448_RestructureAdminAndDelegateTables.cs:70]: The ticket means that a user can have accounts at various centres, and this email address is specific to this centre. It doesn't mean that it should be different from all other account emails at the centre.

File: DigitalLearningSolutions.Data.Migrations/202203161448_RestructureAdminAndDelegateTables.cs [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+            Rename.Column("TCAgreed").OnTable("AdminAccounts").To("TCAgreed_deprecated");
+            Rename.Column("FailedLoginCount").OnTable("AdminAccounts").To("FailedLoginCount_deprecated");
+            Rename.Column("ProfileImage").OnTable("AdminAccounts").To("ProfileImage_deprecated");
+            Rename.Column("SkypeHandle").OnTable("AdminAccounts").To("SkypeHandle_deprecated");
+            Rename.Column("PublicSkypeLink").OnTable("AdminAccounts").To("PublicSkypeLink_deprecated");
+            Rename.Column("ResetPasswordID").OnTable("AdminAccounts").To("ResetPasswordID_deprecated");
+
+            Rename.Table("Candidates").To("DelegateAccounts");
+            Alter.Table("DelegateAccounts").AddColumn("UserID").AsInt32().Nullable().ForeignKey("Users", "ID");
+            Alter.Column("DateRegistered").OnTable("DelegateAccounts").AsDateTime().Nullable();
```
[the remark this answers is not in the document]
Comment (SteveJacksonSoft) [reply DigitalLearningSolutions.Data.Migrations/202203161448_RestructureAdminAndDelegateTables.cs:50]: Sorry, my bad. It should be left non-null.

File: DigitalLearningSolutions.Data.Migrations/202203161448_RestructureAdminAndDelegateTables.cs [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+                .Ascending()
+                .OnColumn("DateRegistered").Ascending().WithOptions().NonClustered();
+            Create.Index("IX_Candidates_CentreID").OnTable("DelegateAccounts").OnColumn("CentreID").Ascending()
+                .WithOptions().NonClustered();
+
+            Rename.Column("EmailAddress").OnTable("DelegateAccounts").To("Email");
+            Alter.Table("DelegateAccounts").AddColumn("EmailVerified").AsDateTime().Nullable();
+            Alter.Column("Answer1").OnTable("DelegateAccounts").AsString(100).Nullable();
+            Alter.Column("Answer2").OnTable("DelegateAccounts").AsString(100).Nullable();
+            Alter.Column("Answer3").OnTable("DelegateAccounts").AsString(100).Nullable();
```
Comment (SteveJacksonSoft) [inline DigitalLearningSolutions.Data.Migrations/202203161448_RestructureAdminAndDelegateTables.cs:74]: What's changing here?

File: DigitalLearningSolutions.Data.Migrations/Scripts/UAR-831-CreateViewsForAdminUsersAndCandidatesTables-UP.sql (hunk 6 above)
Comment (SteveJacksonSoft) [reply DigitalLearningSolutions.Data.Migrations/Scripts/UAR-831-CreateViewsForAdminUsersAndCandidatesTables-UP.sql:12]: Yep.

File: DigitalLearningSolutions.Data.Migrations/Scripts/UAR-831-CreateViewsForAdminUsersAndCandidatesTables-UP.sql [hunk 10]
```diff
@@ -0,0 +1,80 @@
+﻿CREATE VIEW AdminUsers AS
```
Comment (SteveJacksonSoft) [inline DigitalLearningSolutions.Data.Migrations/Scripts/UAR-831-CreateViewsForAdminUsersAndCandidatesTables-UP.sql:1]: We shouldn't have any deprecated columns in here. These views are going to be used by live queries, so if they're pulling in deprecated data we'll get bugs.
We should either use new columns or hard-code values.

Comment (SteveJacksonSoft) [reply DigitalLearningSolutions.Data.Migrations/Scripts/UAR-831-CreateViewsForAdminUsersAndCandidatesTables-UP.sql:1]: Hard-code null or empty values for the columns we're removing.

File: DigitalLearningSolutions.Data.Migrations/Scripts/UAR-831-CreateViewsForAdminUsersAndCandidatesTables-UP.sql [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+       dbo.Users.LastName,
+       dbo.DelegateAccounts.DateRegistered,
+       dbo.DelegateAccounts.CandidateNumber,
+       dbo.Users.JobGroupID,
+       dbo.DelegateAccounts.Answer1,
+       dbo.DelegateAccounts.Answer2,
+       dbo.DelegateAccounts.Answer3,
+       dbo.DelegateAccounts.AliasID_deprecated         AS AliasID,
+       dbo.DelegateAccounts.Approved,
+       dbo.DelegateAccounts.Email                      AS EmailAddress,
```
Comment (SteveJacksonSoft) [inline DigitalLearningSolutions.Data.Migrations/Scripts/UAR-831-CreateViewsForAdminUsersAndCandidatesTables-UP.sql:59]: You want Users.Email as Dan says.
We'll leave admin and delegate emails unused until later in the UAR work when we implement them for simplicity.

File: DigitalLearningSolutions.Data.Migrations/Scripts/UAR-831-CreateViewsForAdminUsersAndCandidatesTables-UP.sql [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+       dbo.AdminAccounts.CentreID,
+       dbo.AdminAccounts.IsCentreAdmin                   AS CentreAdmin,
+       dbo.AdminAccounts.ConfigAdmin_deprecated          AS ConfigAdmin,
+       dbo.AdminAccounts.IsReportsViewer                 AS SummaryReports,
+       dbo.AdminAccounts.IsSuperAdmin                    AS UserAdmin,
+       dbo.Users.FirstName                               AS Forename,
+       dbo.Users.LastName                                AS Surname,
+       dbo.AdminAccounts.Email,
+       dbo.AdminAccounts.IsCentreManager,
+       dbo.AdminAccounts.Approved_deprecated             AS Approved,
```
Comment (SteveJacksonSoft) [inline DigitalLearningSolutions.Data.Migrations/Scripts/UAR-831-CreateViewsForAdminUsersAndCandidatesTables-UP.sql:14]: Hardcode to true

File: DigitalLearningSolutions.Data.Migrations/Scripts/UAR-831-CreateViewsForAdminUsersAndCandidatesTables-UP.sql (hunk 4 above)
Comment (SteveJacksonSoft) [reply DigitalLearningSolutions.Data.Migrations/Scripts/UAR-831-CreateViewsForAdminUsersAndCandidatesTables-UP.sql:2]: Yep

Comment (livzorn) [reply DigitalLearningSolutions.Data.Migrations/Scripts/UAR-831-CreateViewsForAdminUsersAndCandidatesTables-UP.sql]: changed

File: DigitalLearningSolutions.Data.Migrations/202203161448_RestructureAdminAndDelegateTables.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (livzorn) [reply DigitalLearningSolutions.Data.Migrations/202203161448_RestructureAdminAndDelegateTables.cs]: removed this line

File: DigitalLearningSolutions.Data.Migrations/202203161448_RestructureAdminAndDelegateTables.cs [hunk 13]
```diff
@@ -0,0 +1,189 @@
...
+            Delete.Index("IX_Candidates_CandidateNumber").OnTable("DelegateAccounts");
+            Delete.Index("IX_Candidates_CentreID_LastName").OnTable("DelegateAccounts");
+            Delete.Index("IX_Candidates_CentreID_FirstName_LastName").OnTable("DelegateAccounts");
+            Delete.Index("IX_Candidates_Active_CentreID").OnTable("DelegateAccounts");
+            Delete.Index("IX_Candidates_CentreID_DateRegistered").OnTable("DelegateAccounts");
+            Delete.Index("IX_Candidates_CentreID_FirstName").OnTable("DelegateAccounts");
+            Delete.Index("IX_Candidates_CentreID").OnTable("DelegateAccounts");
+            Alter.Column("CandidateNumber").OnTable("DelegateAccounts").AsString(20).NotNullable();
+            Create.Index("IX_Candidates_CandidateNumber").OnTable("DelegateAccounts").OnColumn("CandidateNumber")
+                .Ascending().WithOptions().Unique().WithOptions().NonClustered();
```
[the remark this answers is not in the document]
Comment (livzorn) [reply DigitalLearningSolutions.Data.Migrations/202203161448_RestructureAdminAndDelegateTables.cs:60]: changed

File: DigitalLearningSolutions.Data.Migrations/Scripts/UAR-831-CreateViewsForAdminUsersAndCandidatesTables-UP.sql (hunk 5 above)
Comment (livzorn) [reply DigitalLearningSolutions.Data.Migrations/Scripts/UAR-831-CreateViewsForAdminUsersAndCandidatesTables-UP.sql]: ah cool, removed

File: DigitalLearningSolutions.Data.Migrations/Scripts/UAR-831-CreateViewsForAdminUsersAndCandidatesTables-UP.sql (hunk 6 above)
Comment (livzorn) [reply DigitalLearningSolutions.Data.Migrations/Scripts/UAR-831-CreateViewsForAdminUsersAndCandidatesTables-UP.sql]: changed

File: DigitalLearningSolutions.Data.Migrations/Scripts/UAR-831-CreateViewsForAdminUsersAndCandidatesTables-UP.sql [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+       dbo.AdminAccounts.IsContentManager                AS ContentManager,
+       dbo.AdminAccounts.PublishToAll,
+       dbo.AdminAccounts.ImportOnly,
+       dbo.Users.TermsAgreed                             AS TCAgreed,
+       dbo.AdminAccounts.IsContentCreator                AS ContentCreator,
+       dbo.Users.FailedLoginCount,
+       dbo.Users.ProfileImage,
+       dbo.AdminAccounts.IsSupervisor                    AS Supervisor,
+       dbo.AdminAccounts.IsTrainer                       AS Trainer,
+       dbo.AdminAccounts.CategoryID,
```
[the remark this answers is not in the document]
Comment (livzorn) [reply DigitalLearningSolutions.Data.Migrations/Scripts/UAR-831-CreateViewsForAdminUsersAndCandidatesTables-UP.sql]: Used ISNULL to convert back to 0 if null

File: DigitalLearningSolutions.Data.Migrations/Scripts/UAR-831-CreateViewsForAdminUsersAndCandidatesTables-UP.sql (hunk 10 above)
Comment (livzorn) [reply DigitalLearningSolutions.Data.Migrations/Scripts/UAR-831-CreateViewsForAdminUsersAndCandidatesTables-UP.sql:1]: done

File: DigitalLearningSolutions.Data.Migrations/Scripts/UAR-831-CreateViewsForAdminUsersAndCandidatesTables-UP.sql (hunk 11 above)
Comment (livzorn) [reply DigitalLearningSolutions.Data.Migrations/Scripts/UAR-831-CreateViewsForAdminUsersAndCandidatesTables-UP.sql]: I see, changed

File: DigitalLearningSolutions.Data.Migrations/Scripts/UAR-831-CreateViewsForAdminUsersAndCandidatesTables-UP.sql [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+       dbo.DelegateAccounts.Answer1,
+       dbo.DelegateAccounts.Answer2,
+       dbo.DelegateAccounts.Answer3,
+       dbo.DelegateAccounts.AliasID_deprecated         AS AliasID,
+       dbo.DelegateAccounts.Approved,
+       dbo.DelegateAccounts.Email                      AS EmailAddress,
+       dbo.DelegateAccounts.ExternalReg,
+       dbo.DelegateAccounts.SelfReg,
+       dbo.Users.PasswordHash                          AS Password,
+       dbo.DelegateAccounts.SkipPW_deprecated          AS SkipPW,
```
[the remark this answers is not in the document]
Comment (livzorn) [reply DigitalLearningSolutions.Data.Migrations/Scripts/UAR-831-CreateViewsForAdminUsersAndCandidatesTables-UP.sql]: done

File: DigitalLearningSolutions.Data.Migrations/Scripts/UAR-831-CreateViewsForAdminUsersAndCandidatesTables-UP.sql (hunk 12 above)
Comment (livzorn) [reply DigitalLearningSolutions.Data.Migrations/Scripts/UAR-831-CreateViewsForAdminUsersAndCandidatesTables-UP.sql]: done

File: DigitalLearningSolutions.Data.Migrations/202203161448_RestructureAdminAndDelegateTables.cs (hunk 3 above)
Comment (DanBloxham-sw) [reply DigitalLearningSolutions.Data.Migrations/202203161448_RestructureAdminAndDelegateTables.cs:52]: Okay, that makes sense. Looking at the test DB, candidates has a total of 10 indices. It sounds to me like each of these should be modified in some way. Either we drop them if their relevant details are going to be on the Users table, or we drop them and then recreate them again (to correct their table name in their names).

But it looks like 3 have been left untouched (plus an additional that is now untouched with the latest commit):

- IX_Candidates_CentreID_DateRegistered (see comment below)
- IX_Candidates_Active_CentreID_LastName (should probably be deleted)
- IX_Candidates_CentreID_EmailAddress (should probably be renamed)
- IX_Candidates_CentreIDAliasID (should probably be deleted)

File: DigitalLearningSolutions.Data.Migrations/Scripts/UAR-831-CreateViewsForAdminUsersAndCandidatesTables-UP.sql [hunk 16]
```diff
@@ -54,27 +52,26 @@ SELECT dbo.DelegateAccounts.CandidateID,
...
        dbo.Users.ResetPasswordID,
        dbo.Users.HasBeenPromptedForPrn,
        dbo.Users.ProfessionalRegistrationNumber,
        dbo.Users.LearningHubAuthId,
        dbo.Users.HasDismissedLhLoginWarning
 FROM dbo.Users
-         INNER JOIN
-     dbo.DelegateAccounts ON dbo.Users.ID = dbo.DelegateAccounts.UserID
+         INNER JOIN dbo.DelegateAccounts ON dbo.Users.ID = dbo.DelegateAccounts.UserID
 WHERE dbo.DelegateAccounts.UserID <> NULL
```
Comment (DanBloxham-sw) [inline DigitalLearningSolutions.Data.Migrations/Scripts/UAR-831-CreateViewsForAdminUsersAndCandidatesTables-UP.sql:76]: This WHERE can also be removed.

File: DigitalLearningSolutions.Data.Migrations/Scripts/UAR-831-CreateViewsForAdminUsersAndCandidatesTables-UP.sql [hunk 17]
```diff
@@ -1,45 +1,43 @@
...
        dbo.Users.FailedLoginCount,
        dbo.Users.ProfileImage,
-       dbo.AdminAccounts.IsSupervisor                    AS Supervisor,
-       dbo.AdminAccounts.IsTrainer                       AS Trainer,
-       dbo.AdminAccounts.CategoryID,
-       dbo.AdminAccounts.SkypeHandle_deprecated          AS SkypeHandle,
-       dbo.AdminAccounts.PublicSkypeLink_deprecated      AS PublicSkypeLink,
+       dbo.AdminAccounts.IsSupervisor          AS Supervisor,
+       dbo.AdminAccounts.IsTrainer             AS Trainer,
+       ISNULL(dbo.AdminAccounts.CategoryID, 0),
```
Comment (DanBloxham-sw) [inline DigitalLearningSolutions.Data.Migrations/Scripts/UAR-831-CreateViewsForAdminUsersAndCandidatesTables-UP.sql:29]: I think this is going to need to be aliased as "AS CategoryID" as the ISNULL does not automatically name its output.

File: DigitalLearningSolutions.Data.Migrations/Scripts/UAR-831-CreateViewsForAdminUsersAndCandidatesTables-UP.sql [hunk 18]
```diff
@@ -1,45 +1,43 @@
...
+       dbo.Users.PrimaryEmail                  AS Email,
        dbo.AdminAccounts.IsCentreManager,
-       dbo.AdminAccounts.Approved_deprecated             AS Approved,
-       dbo.AdminAccounts.PasswordReminder_deprecated     AS PasswordReminder,
-       dbo.AdminAccounts.EITSProfile_deprecated          AS EITSProfile,
-       dbo.AdminAccounts.PasswordReminderHash_deprecated AS PasswordReminderHash,
-       dbo.AdminAccounts.PasswordReminderDate_deprecated AS PasswordReminderDate,
+       true                                    AS Approved,
+       0                                       AS PasswordReminder,
+       ""                                      AS EITSProfile,
```
Comment (DanBloxham-sw) [inline DigitalLearningSolutions.Data.Migrations/Scripts/UAR-831-CreateViewsForAdminUsersAndCandidatesTables-UP.sql:16]: I think these are the wrong type of quote marks as I am getting and IDE error here. Should be `''` for SQL

**Review by DanBloxham-sw — CHANGES_REQUESTED** [review]
Some further changes necessary, and a few things to double check.

File: DigitalLearningSolutions.Data.Migrations/Scripts/UAR-831-CreateViewsForAdminUsersAndCandidatesTables-UP.sql (hunk 17 above)
Comment (livzorn) [reply DigitalLearningSolutions.Data.Migrations/Scripts/UAR-831-CreateViewsForAdminUsersAndCandidatesTables-UP.sql]: Ah right, done :)

File: DigitalLearningSolutions.Data.Migrations/Scripts/UAR-831-CreateViewsForAdminUsersAndCandidatesTables-UP.sql (hunk 16 above)
Comment (livzorn) [reply DigitalLearningSolutions.Data.Migrations/Scripts/UAR-831-CreateViewsForAdminUsersAndCandidatesTables-UP.sql]: gone

File: DigitalLearningSolutions.Data.Migrations/202203161448_RestructureAdminAndDelegateTables.cs [hunk 19]
```diff
@@ -44,27 +44,24 @@ public override void Up()
...
 
             Rename.Table("Candidates").To("DelegateAccounts");
             Alter.Table("DelegateAccounts").AddColumn("UserID").AsInt32().Nullable().ForeignKey("Users", "ID");
-            Alter.Column("DateRegistered").OnTable("DelegateAccounts").AsDateTime().Nullable();
 
             Delete.Index("IX_Candidates_CandidateNumber").OnTable("DelegateAccounts");
             Delete.Index("IX_Candidates_CentreID_LastName").OnTable("DelegateAccounts");
             Delete.Index("IX_Candidates_CentreID_FirstName_LastName").OnTable("DelegateAccounts");
             Delete.Index("IX_Candidates_Active_CentreID").OnTable("DelegateAccounts");
-            Delete.Index("IX_Candidates_CentreID_DateRegistered").OnTable("DelegateAccounts");
```
[the remark this answers is not in the document]
Comment (livzorn) [reply DigitalLearningSolutions.Data.Migrations/202203161448_RestructureAdminAndDelegateTables.cs]: Sorry my bad, should not have been removed.

File: DigitalLearningSolutions.Data.Migrations/202203161448_RestructureAdminAndDelegateTables.cs [hunk 20]
```diff
@@ -44,27 +44,24 @@ public override void Up()
             Rename.Column("SkypeHandle").OnTable("AdminAccounts").To("SkypeHandle_deprecated");
             Rename.Column("PublicSkypeLink").OnTable("AdminAccounts").To("PublicSkypeLink_deprecated");
             Rename.Column("ResetPasswordID").OnTable("AdminAccounts").To("ResetPasswordID_deprecated");
+            Delete.ForeignKey("FK_AdminUsers_ResetPasswordID_ResetPassword_ID").OnTable("AdminAccounts");
 
             Rename.Table("Candidates").To("DelegateAccounts");
             Alter.Table("DelegateAccounts").AddColumn("UserID").AsInt32().Nullable().ForeignKey("Users", "ID");
-            Alter.Column("DateRegistered").OnTable("DelegateAccounts").AsDateTime().Nullable();
```
[the remark this answers is not in the document]
Comment (livzorn) [reply DigitalLearningSolutions.Data.Migrations/202203161448_RestructureAdminAndDelegateTables.cs]: Correct!

File: DigitalLearningSolutions.Data.Migrations/202203161448_RestructureAdminAndDelegateTables.cs (hunk 3 above)
Comment (livzorn) [reply DigitalLearningSolutions.Data.Migrations/202203161448_RestructureAdminAndDelegateTables.cs:59]: I've made sure all the foreign keys and indexed on these tables have been renamed correctly. We won't rename keys and indexes on other tables that reference these tables, but it's easy enough to change them here. Those based on deprecated columns have been removed, and the rest have been updated to reflect the new table names.

**Comment by SteveJacksonSoft** [comment]
Can we hold off on merging this until we're happy the UAR env and jenkins pipeline are working right? I'm quite keen to avoid accidentally running this on Test.

File: DigitalLearningSolutions.Data.Migrations/202203161448_RestructureAdminAndDelegateTables.cs [hunk 21]
```diff
@@ -53,23 +60,41 @@ public override void Up()
...
             Alter.Column("Answer1").OnTable("DelegateAccounts").AsString(100).Nullable();
             Alter.Column("Answer2").OnTable("DelegateAccounts").AsString(100).Nullable();
             Alter.Column("Answer3").OnTable("DelegateAccounts").AsString(100).Nullable();
 
+            Delete.ForeignKey("FK_Candidates_Centres").OnTable("DelegateAccounts");
+            Create.ForeignKey("FK_DelegateAccounts_Centres").FromTable("DelegateAccounts").ForeignColumn("CentreID")
+                .ToTable("Centres").PrimaryColumn("CentreID");
+            Delete.ForeignKey("FK_Candidates_JobGroups").OnTable("DelegateAccounts");
+            Create.ForeignKey("FK_DelegateAccounts_JobGroups").FromTable("DelegateAccounts").ForeignColumn("JobGroupID")
+                .ToTable("JobGroups").PrimaryColumn("JobGroupID");
```
Comment (DanBloxham-sw) [inline DigitalLearningSolutions.Data.Migrations/202203161448_RestructureAdminAndDelegateTables.cs:96]: JobGroup is going in the new Users table, so I don't think this new creation is necessary.

**Review by DanBloxham-sw — CHANGES_REQUESTED** [review]
Just one foreign key that I don't think needs to be created, otherwise I think this looks good.

File: DigitalLearningSolutions.Data.Migrations/202203161448_RestructureAdminAndDelegateTables.cs (hunk 21 above)
Comment (livzorn) [reply DigitalLearningSolutions.Data.Migrations/202203161448_RestructureAdminAndDelegateTables.cs]: Thanks for reminding me, I've removed the recreation

**Review by DanBloxham-sw — APPROVED** [review]
Looks good. Just need to wait on merging as Steve's comment says.

**Comment by SteveJacksonSoft** [comment]
@livzorn Sorry, a couple of new requirements now that I've thought through the data migration more.
- Can we add a datetime column to CandidateAccounts called "DetailsLastChecked"? We've got an upcoming requirement to prompt users to update their details every 6 months which we'll use this for, and it'll also be useful for getting users to resolve conflicts after the data migration. On CandidateAccounts because this gives us finer control than on Users and some details are centre-specific.
- Can we rename the CandidateAccounts.Password field to "OldPassword" instead? This is for handling the data transition when we have multiple candidates which have the same email but different password hashes. When a user logs in after the refactor, we'll compare the password they give us with the "old passwords" so we can see which delegate accounts to let them into and tell them if we need them to reset their password so they can get into the other delegate accounts matching their email.
- It's possible that JobGroup will be moving back to CandidateAccounts or becoming nullable, but I need to check with Kevin first.

**Comment by livzorn** [comment]
Update: JobGroup is staying on Users and staying not-null

File: DigitalLearningSolutions.Data.Migrations/Scripts/UAR-831-CreateViewsForAdminUsersAndCandidatesTables-UP.sql [hunk 22]
```diff
@@ -0,0 +1,76 @@
...
+       dbo.Users.JobGroupID,
+       dbo.DelegateAccounts.Answer1,
+       dbo.DelegateAccounts.Answer2,
+       dbo.DelegateAccounts.Answer3,
+       null                          AS AliasID,
+       dbo.DelegateAccounts.Approved AS Approved,
+       dbo.Users.PrimaryEmail        AS EmailAddress,
+       dbo.DelegateAccounts.ExternalReg,
+       dbo.DelegateAccounts.SelfReg,
+       dbo.Users.PasswordHash        AS Password,
```
Comment (DanBloxham-sw) [inline DigitalLearningSolutions.Data.Migrations/Scripts/UAR-831-CreateViewsForAdminUsersAndCandidatesTables-UP.sql:60]: With the change of `Password` to `OldPassword` on `DelegateAccounts`, should we be returning that here, or are we keeping the `Users.PasswordHash` here, and making new functions later on for the specific login case?

**Review by DanBloxham-sw — COMMENTED** [review]
One question about we plan to use the renamed password column. Otherwise looks good.

File: DigitalLearningSolutions.Data.Migrations/Scripts/UAR-831-CreateViewsForAdminUsersAndCandidatesTables-UP.sql (hunk 22 above)
Comment (SteveJacksonSoft) [reply DigitalLearningSolutions.Data.Migrations/Scripts/UAR-831-CreateViewsForAdminUsersAndCandidatesTables-UP.sql:60]: We'll keep Users.PasswordHash here and make new functions later on.

**Review by DanBloxham-sw — APPROVED** [review]
Looks all good then. Again, don't know if it is ready to merge with the environments and all that.

**Pull request merged**